File: php/src/model/mxGeometry.php
<?php
/**
 * $Id: mxGeometry.php,v 1.18 2010-06-29 12:05:42 gaudenz Exp $
 * Copyright (c) 2006-2010, Gaudenz Alder
 */
class mxGeometry extends mxRectangle
{

	/**
	 * Class: mxGeometry
	 *
	 * Represents the geometry of a cell. For vertices, the geometry consists
	 * of the x- and y-location, as well as the width and height. For edges,
	 * the edge either defines the source- and target-terminal, or the geometry
	 * defines the respective terminal points.
	 * 
	 * Variable: TRANSLATE_CONTROL_POINTS
	 * 
	 * Global switch to translate the points in translate. Default is true.
	 */
	public static $TRANSLATE_CONTROL_POINTS = true;

	/**
	 * Variable: alternateBounds
	 *
	 * Stores alternate values for x, y, width and height in a rectangle.
	 * Default is null.
	 */
	var $alternateBounds;

	/**
	 * Variable: sourcePoint
	 *
	 * Defines the source point of the edge. This is used if the corresponding
	 * edge does not have a source vertex. Otherwise it is ignored. Default is
	 * null.
	 */
	var $sourcePoint;
	
	/**
	 * Variable: targetPoint
	 *
	 * Defines the target point of the edge. This is used if the corresponding
	 * edge does not have a target vertex. Otherwise it is ignored. Default is
	 * null.
	 */
	var $targetPoint;
		
	/**
	 * Variable: points
	 *
	 * Array of <mxPoints> which specifies the control points along the edge.
	 * These points are the intermediate points on the edge, for the endpoints
	 * use <targetPoint> and <sourcePoint> or set the terminals of the edge to
	 * a non-null value. Default is null.
	 */
	var $points;
	
	/**
	 * Variable: offset
	 *
	 * Holds the offset of the label for edges. This is the absolute vector
	 * between the center of the edge and the top, left point of the label.
	 * Default is null.
	 */
	var $offset;

	/**
	 * Variable: relative
	 *
	 * Specifies if the coordinates in the geometry are to be interpreted as
	 * relative coordinates. Default is false. This is used to mark a geometry
	 * with an x- and y-coordinate that is used to describe an edge label
	 * position.
	 */
	var $relative = false;

	/**
	 * Constructor: mxGeometry
	 *
	 * Constructs a new object to describe the size and location
	 * of a vertex or the control points of an edge.
	 */
	function mxGeometry($x=0, $y=0, $width=0, $height=0)
	{
		parent::mxRectangle($x, $y, $width, $height);
	}

	/**
	 * Function: getTerminalPoint
	 * 
	 * Returns the <mxPoint> representing the source or target point of this
	 * edge. This is only used if the edge has no source or target vertex.
	 * 
	 * Parameters:
	 * 
	 * isSource - Boolean that specifies if the source or target point
	 * should be returned.
	 */
	function getTerminalPoint($isSource)
	{
		return ($isSource) ? $this->sourcePoint : $this->targetPoint;
	}
	
	/**
	 * Function: setTerminalPoint
	 * 
	 * Sets the <sourcePoint> or <targetPoint> to the given <mxPoint> and
	 * returns the new point.
	 * 
	 * Parameters:
	 * 
	 * point - Point to be used as the new source or target point.
	 * isSource - Boolean that specifies if the source or target point
	 * should be set.
	 */
	function setTerminalPoint($point, $isSource)
	{
		if ($isSource)
		{
			$this->sourcePoint = $point;
		}
		else
		{
			$this->targetPoint = $point;
		}
		
		return $point;
	}
	
	/**
	 * Function: translate
	 * 
	 * Translates the geometry by the specified amount. That is, <x> and <y>
	 * of the geometry, the <sourcePoint>, <targetPoint> and all elements of
	 * <points> are translated by the given amount. <x> and <y> are only
	 * translated if <relative> is false. If <TRANSLATE_CONTROL_POINTS> is
	 * false, then <points> are not modified by this function.
	 * 
	 * Parameters:
	 * 
	 * dx - Integer that specifies the x-coordinate of the translation.
	 * dy - Integer that specifies the y-coordinate of the translation.
	 */
	function translate($dx, $dy)
	{
		// Translates the geometry
		if (!$this->relative)
		{
			$this->x += $dx;
			$this->y += $dy;
		}
		
		// Translates the source point
		if ($this->sourcePoint != null)
		{
			$this->sourcePoint->x += $dx;
			$this->sourcePoint->y += $dy;
		}
		
		// Translates the target point
		if ($this->targetPoint != null)
		{
			$this->targetPoint->x += $dx;
			$this->targetPoint->y += $dy;
		}

		// Translate the control points
		if (mxGeometry::$TRANSLATE_CONTROL_POINTS &&
			$this->points != null)
		{
			$count = sizeof($this->points);
			
			for ($i = 0; $i < $count; $i++)
			{
				$pt = $this->points[i];
				
				$pt->x += $dx;
				$pt->y += $dy;
			}
		}
	}
	
	/**
	 * Function: copy
	 *
	 * Returns a copy of this <mxGeometry>.
	 */
	function copy()
	{
	 	$clone = new mxGeometry($this->x, $this->y, $this->width, $this->height);
				
		// Clones the points
		if ($this->points != null)
		{
			$clone->points = array();
			
			for ($i = 0; $i < sizeof($this->points); $i++)
			{
				array_push($clone->points, $this->points[$i]->copy());
			}
		}
		
		// Clones the alternatebounds
		if ($this->alternateBounds != null)
		{
			$clone->alternateBounds = $this->alternateBounds->copy();
		}
		
		// Clones the offset
		if ($this->offset != null)
		{
			$clone->offset = $this->offset->copy();
		}
		
		// Clones the source and targetpoint
		if ($this->sourcePoint != null)
		{
			$clone->sourcePoint = $this->sourcePoint->copy();
		}
		
		if ($this->targetPoint != null)
		{
			$clone->targetPoint = $this->targetPoint->copy();
		}

	 	$clone->relative = $this->relative;

		return $clone;
	}

}

?>

File: dotnet/src/model/mxGeometry.cs
// Copyright (c) 2007-2008, Gaudenz Alder
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;
using System.Drawing;

namespace com.mxgraph
{
    /// <summary>
    /// Represents the geometry of a cell. For vertices, the geometry consists
    /// of the x- and y-location, as well as the width and height. For edges,
    /// the edge either defines the source- and target-terminal, or the geometry
    /// defines the respective terminal points.
    /// </summary>
    public class mxGeometry : mxRectangle
    {

        /// <summary>
        /// Global switch to translate the points in translate. Default is true.
        /// </summary>
        public static bool TRANSLATE_CONTROL_POINTS = true;

        /// <summary>
        /// Stores alternate values for x, y, width and height in a rectangle.
        /// Default is null.
        /// </summary>
        protected mxRectangle alternateBounds;

        /// <summary>
        /// Defines the source-point of the edge. This is used if the
        /// corresponding edge does not have a source vertex. Otherwise it is
        /// ignored. Default is null.
        /// </summary>
        protected mxPoint sourcePoint;

        /// <summary>
        /// Defines the target-point of the edge. This is used if the
        /// corresponding edge does not have a source vertex. Otherwise it is
        /// ignored. Default is null.
        /// </summary>
        protected mxPoint targetPoint;

        /// <summary>
        /// Holds the offset of the label for edges. This is the absolute vector
        /// between the center of the edge and the top, left point of the label.
        /// Default is null.
        /// </summary>
        protected mxPoint offset;

        /// <summary>
        /// List of mxPoints which specifies the control points along the edge.
        /// These points are the intermediate points on the edge, for the endpoints
        /// use targetPoint and sourcePoint or set the terminals of the edge to
        /// a non-null value. Default is null.
        /// </summary>
        protected List<mxPoint> points;

        /// <summary>
        /// Specifies if the coordinates in the geometry are to be interpreted as
        /// relative coordinates. Default is false. This is used to mark a geometry
        /// with an x- and y-coordinate that is used to describe an edge label
        /// position.
        /// </summary>
        protected bool relative = false;

        /// <summary>
        /// Constructs a new geometry at (0, 0) with the width and height set to 0.
        /// </summary>
        public mxGeometry() : this(0, 0, 0, 0) { }

        /// <summary>
        /// Constructs a geometry using the given parameters.
        /// </summary>
        /// <param name="x">X-coordinate of the new geometry.</param>
        /// <param name="y">Y-coordinate of the new geometry.</param>
        /// <param name="width">Width of the new geometry.</param>
        /// <param name="height">Height of the new geometry.</param>
        public mxGeometry(double x, double y, double width, double height) : base(x, y, width, height) { }

        /// <summary>
        /// Constructs a copy of the given geometry.
        /// </summary>
        /// <param name="geometry">Geometry to construct a copy of.</param>
        public mxGeometry(mxGeometry geometry)
            : base(geometry.X, geometry.Y, geometry.Width, geometry
                    .Height)
        {
            if (geometry.points != null)
            {
                points = new List<mxPoint>(geometry.points.Count);

                foreach (mxPoint pt in geometry.points)
                {
                    points.Add(pt.Clone());
                }
            }

            if (geometry.sourcePoint != null)
            {
                sourcePoint = geometry.sourcePoint.Clone();
            }

            if (geometry.targetPoint != null)
            {
                targetPoint = geometry.targetPoint.Clone();
            }

            if (geometry.offset != null)
            {
                offset = geometry.offset.Clone();
            }

            if (geometry.alternateBounds != null)
            {
                alternateBounds = geometry.alternateBounds.Clone();
            }

            relative = geometry.relative;
        }

        /// <summary>
        /// Sets or returns the alternate bounds.
        /// </summary>
        public mxRectangle AlternateBounds
        {
            get { return alternateBounds; }
            set { alternateBounds = value; }
        }

        /// <summary>
        /// Sets or returns the source point.
        /// </summary>
        public mxPoint SourcePoint
        {
            get { return sourcePoint; }
            set { sourcePoint = value; }
        }

        /// <summary>
        /// Sets or returns the target point.
        /// </summary>
        public mxPoint TargetPoint
        {
            get { return targetPoint; }
            set { targetPoint = value; }
        }

        /// <summary>
        /// Sets or returns the list of control points.
        /// </summary>
        public List<mxPoint> Points
        {
            get { return points; }
            set { points = value; }
        }

        /// <summary>
        /// Sets or returns the offset.
        /// </summary>
        public mxPoint Offset
        {
            get { return offset; }
            set { offset = value; }
        }

        /// <summary>
        /// Sets or returns if the geometry is relative.
        /// </summary>
        public bool Relative
        {
            get { return relative; }
            set { relative = value; }
        }

        /// <summary>
        /// Returns the point representing the source or target point of this edge.
        /// This is only used if the edge has no source or target vertex.
        /// </summary>
        /// <param name="source">Boolean that specifies if the source or target point
        /// should be returned.</param>
        /// <returns>Returns the source or target point.</returns>
        public mxPoint GetTerminalPoint(bool source)
        {
            return (source) ? sourcePoint : targetPoint;
        }

        /// <summary>
        /// Sets the sourcePoint or targetPoint to the given point and returns the
        /// new point.
        /// </summary>
        /// <param name="point">Point to be used as the new source or target point.</param>
        /// <param name="source">Boolean that specifies if the source or target point
        /// should be set.</param>
        /// <returns>Returns the new point.</returns>
        public mxPoint SetTerminalPoint(mxPoint point, bool source)
        {
            if (source)
            {
                sourcePoint = point;
            }
            else
            {
                targetPoint = point;
            }

            return point;
        }

        /// <summary>
        /// Translates the geometry by the specified amount. That is, x and y of the
        /// geometry, the sourcePoint, targetPoint and all elements of points are
        /// translated by the given amount. X and y are only translated if the
        /// geometry is not relative. If TRANSLATE_CONTROL_POINTS is false, then
        /// are not modified by this function.
        /// </summary>
        /// <param name="dx">Integer that specifies the x-coordinate of the translation.</param>
        /// <param name="dy">Integer that specifies the y-coordinate of the translation.</param>
        public void Translate(double dx, double dy)
        {
            // Translates the geometry
            if (!Relative)
            {
                x += dx;
                y += dy;
            }

            // Translates the source point
            if (sourcePoint != null)
            {
                sourcePoint.X += dx;
                sourcePoint.Y += dy;
            }

            // Translates the target point
            if (targetPoint != null)
            {
                targetPoint.X += dx;
                targetPoint.Y += dy;
            }

            // Translate the control points
            if (TRANSLATE_CONTROL_POINTS &&
                points != null)
            {
                int count = points.Count;

                for (int i = 0; i < count; i++)
                {
                    mxPoint pt = points[i];

                    pt.X += dx;
                    pt.Y += dy;
                }
            }
        }

        /// <summary>
        /// Returns a new instance of the same geometry.
        /// </summary>
        /// <returns>Returns a clone of the geometry.</returns>
        public new mxGeometry Clone()
        {
            return new mxGeometry(this);
        }

    }

}
